File: generated_code/python/solution_1411_2.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def getDecimalValue(self, head: Optional[ListNode]) -> int:
        # Time Complexity: O(N) where N is the number of nodes in the linked list.
        # Space Complexity: O(1) as we only use a single variable to accumulate the result.
        
        ans = 0
        current = head
        
        while current:
            # Shift the existing bits to the left by 1 (equivalent to multiplying by 2)
            # and add the current node's value using the bitwise OR operator.
            ans = (ans << 1) | current.val
            current = current.next
            
        return ans

File: generated_code/cpp/solution_1411_2.cpp
#include <iostream>

// Definition for singly-linked list, included to ensure code is self-contained.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    int getDecimalValue(ListNode* head) {
        int result = 0;
        
        // Traverse the linked list from head to tail.
        // Since the head represents the most significant bit, we shift the 
        // accumulated result to the left by 1 (equivalent to multiplying by 2) 
        // and add the current node's value.
        // Time Complexity: O(N) where N is the number of nodes.
        // Space Complexity: O(1) auxiliary space.
        while (head != nullptr) {
            result = (result << 1) | head->val;
            head = head->next;
        }
        
        return result;
    }
};